GUI/GUIListPanel.cpp

```diff
@@ -824,6 +838,88 @@ void GUIListPanel::ScrollToBottom()
 }
 
 
+/// <summary>
+/// Sets whether the scroll panel scrolls in a loop or not.
+/// </summary>
+/// <param name="scrollLoop">true to scroll in a loop, false to scroll with edge stopping.</param>
+void GUIListPanel::SetMouseScrollingLoop(bool ScrollLoop)
+{
+	m_LoopMouseScroll = ScrollLoop;
+}
+
+
+/// <summary>
+/// Sets whether the list panel can be scrolled with the mouse scroll wheel.
+/// </summary>
+/// <param name="MouseScroll">true to enable scrolling, false to disable</param>
+void GUIListPanel::SetMouseScrolling(bool MouseScroll)
+{
+	m_MouseScroll = MouseScroll;
+}
+
+
+/// <summary>
+/// Perform list scrolling through the scrollbar
+/// </summary>
+/// <param name="MouseWheelChange">amount and direction of scrolling. positive to scroll up, negative to scroll down.</param>
+void GUIListPanel::ScrollBarScrolling(int MouseWheelChange)
+{
+	int newValue = 0;
+	Item* lastItem = GetItem(GetItemList()->size() - 1);
+	int avgItemHeight = static_cast<int>((GetStackHeight(lastItem) + GetItemHeight(lastItem)) / GetItemList()->size());
+	if (MouseWheelChange < 0) {
+		newValue = m_VertScroll->GetValue() - (MouseWheelChange * avgItemHeight);
+		int maxValue = GetStackHeight(lastItem) + GetItemHeight(lastItem) - m_VertScroll->GetPageSize();
+		if (maxValue < 0) {
+			newValue = 0;
+		} else if (newValue > maxValue) {
+			newValue = maxValue;
+		}
+	} else {
+		newValue = m_VertScroll->GetValue() - (MouseWheelChange * avgItemHeight);
+		if (newValue < 0) {
+			newValue = 0;
+		}
+	}
+	m_VertScroll->SetValue(newValue);
+	BuildBitmap(false, true);
+}
+
+
+/// <summary>
+/// Perform list scrolling by changing the currently selected list item.
+/// Should not be used with multiselect enabled.
+/// </summary>
+/// <param name="ScrollDir">Direction of scrolling. positive to scroll up, negative to scroll down.</param>
+void GUIListPanel::SelectionListScrolling(int ScrollDir)
+{
+	int newItemIndex = 0;
+	int OldItemIndex = GetSelectedIndex();
+	if (ScrollDir > 0) {
```

Comment:
This stuff can probably be simplified to use scrollDir directly, assuming it's limited to -1 and 1, and generally trimmed down. i.e.
```
if (scrollDir == 1 || scrollDir == -1) { //maybe this is unnecessary
	std::size_t itemListSize = GetItemList()->size();

	int newItemIndex = GetSelectedIndex() + scrollDir; //feel free to make a local variable for GetSelectedIndex if you want, I just didn't bother
	if (newItemIndex < 0 || newItemIndex >= itemListSize) {
		if (m_LoopMouseScroll) {
			newItemIndex = (newItemIndex + itemListSize) % itemListSize; //Unless I'm messing up my math, this should solve the wrap-around problem
		} else {
			newItemIndex = Clamp(newItemIndex, itemListSize - 1, 0);
			if (GetSelectedIndex() == newItemIndex) {
				SendSignal(EdgeHit, newItemIndex < 0 ? 0 : 1);
				return;
			}
		}
	}
	SetSelectedIndex(newItemIndex);
}
```

Also, a bit of general stuff; I know I never made a style guide like I said I would (oops) and also the gui code is kind of its own weirdness, so it's not a big deal, but normally method params and local variables use camelCase, and comments should follow English grammar and have capital letters after periods. Like so.

Reply:
thanks, it's definitely shorter now. though specifically about scrollDir, there's no guarantee of it being clamped between -1 and 1 as far as i can tell, so i don't want to rely on that (it depends on allegro's mouse_z variable).
i'll try to pay more attention to writing in that style, i just don't have much experience programming in any style. :)

Reply:
Thanks for moving it. I'd also chuck a TODO in there saying to either implement it or remove (describe the edge case if you can remember it) so we know it's there for a reason.